controllers/cks_cluster_controller.go

```diff
@@ -0,0 +1,124 @@
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackclusters/status,verbs=create;get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackclusters/finalizers,verbs=update
+
+// CksClusterReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackClusters.
+// The runner does the actual reconciliation.
+type CksClusterReconciliationRunner struct {
+	*csCtrlrUtils.ReconciliationRunner
+	FailureDomains        *infrav1.CloudStackFailureDomainList
+	ReconciliationSubject *infrav1.CloudStackCluster
+}
+
+// CksClusterReconciler is the k8s controller manager's interface to reconcile a CloudStackCluster.
+// This is primarily to adapt to k8s.
+type CksClusterReconciler struct {
+	csCtrlrUtils.ReconcilerBase
+}
+
+// Initialize a new CloudStackCluster reconciliation runner with concrete types and initialized member fields.
+func NewCksClusterReconciliationRunner() *CksClusterReconciliationRunner {
+	// Set concrete type and init pointers.
+	runner := &CksClusterReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackCluster{}}
+	runner.FailureDomains = &infrav1.CloudStackFailureDomainList{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	runner.ReconciliationRunner = csCtrlrUtils.NewRunner(runner, runner.ReconciliationSubject, "CKSClusterController")
+	runner.CSCluster = runner.ReconciliationSubject
+	return runner
+}
+
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	r := NewCksClusterReconciliationRunner()
+	r.UsingBaseReconciler(reconciler.ReconcilerBase).ForRequest(req).WithRequestCtx(ctx)
+	r.WithAdditionalCommonStages(r.GetFailureDomains(r.FailureDomains))
+	return r.RunBaseReconciliationStages()
+}
+
+// Reconcile actually reconciles the CloudStackCluster.
+func (r *CksClusterReconciliationRunner) Reconcile() (res ctrl.Result, reterr error) {
+	if r.CSCluster.Spec.SyncWithACS {
+		// Prevent premature deletion.
+		controllerutil.AddFinalizer(r.ReconciliationSubject, CksClusterFinalizer)
+
+		if len(r.FailureDomains.Items) == 0 {
+			return r.RequeueWithMessage("No failure domains found")
+		}
+
+		res, err := r.AsFailureDomainUser(&r.FailureDomains.Items[0].Spec)()
+		if r.ShouldReturn(res, err) {
+			return res, err
+		}
+
+		r.Log.Info("Creating entry with CKS")
+		err = r.CSUser.GetOrCreateCksCluster(r.CAPICluster, r.ReconciliationSubject, &r.FailureDomains.Items[0].Spec)
+		if err != nil {
+			if strings.Contains(err.Error(), "Kubernetes Service plugin is disabled") {
+				r.Log.Info("Kubernetes Service plugin is disabled on CloudStack. Skipping ExternalManaged kubernetes cluster creation")
+				return ctrl.Result{}, nil
+			}
+			r.Log.Info(fmt.Sprintf("Failed creating ExternalManaged kubernetes cluster on CloudStack. Error: %s", err.Error()))
```

Comment:
logr favors using key values instead of string interpolation. also wouldn't controller runtime log this error if we are returning an error?
```suggestion
			r.Log.Info("Failed creating ExternalManaged kubernetes cluster on CloudStack", "error", err.Error()))
```
I think Log also has log.Error